homeassistant/components/rainmachine/__init__.py

```diff
@@ -127,14 +134,20 @@
 
     conf = config[DOMAIN]
 
-    if conf[CONF_IP_ADDRESS] in configured_instances(hass):
-        return True
+    for controller in conf[CONF_CONTROLLERS]:
+        if controller[CONF_IP_ADDRESS] in configured_instances(hass):
+            continue
 
-    hass.async_create_task(
-        hass.config_entries.flow.async_init(
-            DOMAIN,
-            context={'source': SOURCE_IMPORT},
-            data=conf))
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={'source': SOURCE_IMPORT},
+                data={
+                    CONF_IP_ADDRESS: controller[CONF_IP_ADDRESS],
+                    CONF_PASSWORD: controller[CONF_PASSWORD],
+                    CONF_PORT: controller.get(CONF_PORT, DEFAULT_PORT),
```

Comment:
Aren't defaults already in place from the config validation?

Reply:
Defaults are only in place for when the config UI is used; in the case where `configuation.yaml`'s values are imported, there aren't defaults being applied.

Reply:
My `CONFIG_SCHEMA` lists `port` and `ssl` as optional; if they aren't provided in `configuration.yaml`, they don't flow through to `async_setup_entry` correctly. For example, with this `configuration.yaml`:

```
rainmachine:
  controllers:
    - ip_address: !secret rm_host
      password: !secret rm_password
```

...I get this:

```
2018-12-04 11:40:37 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry rainmachine.phil.iot for rainmachine
Traceback (most recent call last):
  File "/Users/abach/Git/home-assistant/homeassistant/config_entries.py", line 249, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/Users/abach/Git/home-assistant/homeassistant/components/rainmachine/__init__.py", line 157, in async_setup_entry
    port = config_entry.data[CONF_PORT]
KeyError: 'port'
```

...because the port is not being delivered.

That said, I think you're right that the config entry should handle this. I'll update.

Reply:
Check out e1df665657da and tell me you like this approach. 👍

Reply:
It looks good!

But I'm wondering why you got the stacktrace above, since as long as `async_setup_component` is used to set up the component, the config validation should happen. Wonder what is not as we expect?

Reply:
It looks like the config validation doesn't actually populate the `config` dict with the defaults. For example, if I don't provide the port, this is what I see:

```
2018-12-04 13:04:28 DEBUG (MainThread) [homeassistant.components.rainmachine] OrderedDict([('controllers', [OrderedDict([('ip_address', '192.168.1.100'), ('password', '12345abcde')])])])
```